pkg/maputil/cmpmap.go

```diff
@@ -0,0 +1,65 @@
+package maputil
+
+///
+// CmpMap is a version of syncmap, which compares the values before inserting.
+//
+
+type cmpMapImpl[K comparable, V any] struct {
+	syncMapImpl[K, V]
+	cmp func(a, b V) bool
+}
+
+// NewCmpMap creates a new comparing map, which only updates the stored values if the
+// cmp predicate returns true when provided with the existing and new values.
+// If cmp is nil, the map works as a normal map, with synchronized access.
+//
+// Example:
+//
+//	m := NewCmpMap[string](Max[int])
+//	m.Add("a", 10)
+//	m.Add("a", 5)
+//	v, ok := m.Get("a") // 10
+func NewCmpMap[K comparable, V any](cmp func(a, b V) bool) SyncMap[K, V] {
+	return &cmpMapImpl[K, V]{
+		syncMapImpl: syncMapImpl[K, V]{data: make(map[K]V)},
+		cmp:         cmp}
+}
+
+// Store inserts the value v to the map at the key k, or updates the value if the
+// comparison predicate returns true.
+func (m *cmpMapImpl[K, V]) Store(k K, v V) {
+	if m.cmp == nil {
+		m.syncMapImpl.Store(k, v)
+		return
+	}
+	m.mux.Lock()
+	defer m.mux.Unlock()
+	if m.data == nil {
+		m.data = make(map[K]V, 1)
+	}
+	if existing, ok := m.data[k]; ok {
+		if !m.cmp(existing, v) {
+			return
+		}
+	}
+	m.data[k] = v
```

Comment:
Why do you need to embed `syncMapImpl` into `cmpMapImpl` and manipulate the mutex if the former provides `Access()` and `RAccess()`?

Reply:
It didn't provide at the time of implementation. Indeed, now it makes sense to use those methods. Thanks.